aten/src/ATen/native/sparse/cuda/SoftMax.cu

```diff
@@ -0,0 +1,640 @@
+__global__ void cuda_sparse_coo_softmax_kernel(
+    int64_t* sorted_pool_indices,
+    int64_t size,
+    int64_t* pool_sizes,
+    int64_t* pool_offsets,
+    int64_t nvalues,
+    scalar_t* mx_rows,
+    PackedTensorAccessor<scalar_t, 2> input_values_acc,
+    PackedTensorAccessor<scalar_t, 2> ouput_values_acc) {
+  /*
+    See ATen/native/sparse/Softmax.cpp:cpu_sparse_coo_softmax for the CPU
+    implementation of the sparse softmax algorithm that this implementation is
+    based on.
+  */
+  int tid = threadIdx.x;
+  int blkid = blockIdx.x;
+  int blksz = blockDim.x;
+  int gridsz = gridDim.x;
+
+  int index = tid + blkid * blksz;
+  int step = blksz * gridsz;
+
+  while (index < size) {
+    int64_t offset = pool_offsets[index];
+    int64_t* pool_indices = sorted_pool_indices + offset;
+    int64_t pool_indices_size = pool_sizes[index];
+    scalar_t* mx_row = mx_rows + index * nvalues;
+
+    for (int64_t j = 0; j < nvalues; j++) {
+      scalar_t exp_sums = 0;
+      for (int64_t p = 0; p < pool_indices_size; p++) {
+        auto i = pool_indices[p];
+        auto values_row = input_values_acc[i];
+        auto out_values_row = ouput_values_acc[i];
+
+        auto v = c10::cuda::compat::exp(values_row[j] - mx_row[j]);
+        if (!LogSoftMax) {
+          out_values_row[j] = v;
+        }
+        exp_sums += v;
+      }
+      for (int64_t p = 0; p < pool_indices_size; p++) {
+        auto i = pool_indices[p];
+        auto values_row = input_values_acc[i];
+        auto out_values_row = ouput_values_acc[i];
+        
+        if (LogSoftMax) {
+          out_values_row[j] = values_row[j] - mx_row[j] - c10::cuda::compat::log(exp_sums);
+        } else {
+          out_values_row[j] *= 1.0 / exp_sums;
+        }
+      }
+    }
+    index += step;
+  }
+}
+
+template <typename scalar_t, bool LogSoftMax>
+__global__ void cuda_sparse_coo_softmax_backward_kernel(
+    int64_t* sorted_pool_indices,
+    int64_t size,
+    int64_t* pool_sizes,
+    int64_t* pool_offsets,
+    int64_t nvalues,
+    int64_t grad_nnz,
+    scalar_t* exp_sums_rows,
+    int64_t* grad_offsets,
+    int64_t* out_offsets,
+    int64_t* lower_bound_values,
+    PackedTensorAccessor<scalar_t, 2> values_accessor,
+    PackedTensorAccessor<scalar_t, 2> out_values_accessor,
+    PackedTensorAccessor<scalar_t, 2> grad_values_accessor) {
+  /*
+    See ATen/native/sparse/Softmax.cpp:cpu_sparse_coo_softmax_backward for
+    the CPU implementation of the sparse softmax backward algorithm that this
+    implementation is based on.
+  */
+  int tid = threadIdx.x;
+  int blkid = blockIdx.x;
+  int blksz = blockDim.x;
+  int gridsz = gridDim.x;
+
+  int index = tid + blkid * blksz;
+  int step = blksz * gridsz;
+
+  while (index < size) {
+    int64_t offset = pool_offsets[index];
+    int64_t* pool_indices = sorted_pool_indices + offset;
+    int64_t pool_indices_size = pool_sizes[index];
+
+    scalar_t* tmp_row = exp_sums_rows + index * nvalues;
```

Comment:
For exp_sums_rows, it looks like:

- each element of exp_sums_rows is only accessed by a single thread
- the value of exp_sums_rows after the kernel is run is irrelevant

This goes back to an earlier question: why is it needed vs. just creating a local array of size nvalues initialized to zero?

Reply:
(Basically: can we port the improvements you made to forward to backward.)

Reply:
All right